### ISSUE TITLE
template-based approach for output generation

### ISSUE THREAD

**Issue by FedeBev** [drwpow/openapi-typescript#499]
Hi there,

Right now, all the outputs are generated with string concatenation, that is fine for simple and easy cases.
Why don't we think of a model-based approach, for example with something like [ejs](https://www.npmjs.com/package/ejs)

In my opinion, the pros would be:
- output format easier to understand
- more maintainable code
- less error-prone
- easier to improve and extend
- easier to add documentation and annotations, such as `@deprecated`

Thanks

**Comment by drwpow** [comment]
Thanks for the suggestion! 

When you propose a “model-based approach,” the hidden complexity is determining what that model is. And that’s something I’ve tried to intentionally avoid in this library.

When I wrote v2, it was a complete rewrite. I tried to get away from string concatenation (which was much messier than it is now), and move towards a JSON-parsable format. But my problem was that OpenAPI already is a JSON-parsable format. And TypeScript does not map 1:1 to a JSON format at all. So I quickly ran into the barriers of creating some weird, intermediary format between OpenAPI and TypeScript, that was neither. It was just an unnecessary abstraction that complicated everything.

But to your suggestion, one thing we could probably do is convert OpenAPI -> TypeScript AST, which would then automatically compile to TypeScript. While this wouldn’t use a template, it would preserve your “model” idea. I’m not opposed to this approach; I just simply saw string concatenation as easier to deal with. And it would make some things harder. For example, we convert `description`s into JSDoc comments. That involves hoisting `description` above to its parent context, and inserting a sibling node before the original schema. Just finding the right place to create & insert a new node in an AST can be frustrating. But in string concatenation, it’s trivial: just put it before the schema.

Anyways, if I’ve missed your intention entirely I apologize. I’d be open to a proof-of-concept if you feel there’s a clear win here, but with all the testing we have, I’m not convinced a model/template approach would be easier to maintain or less error prone; I’d have to see a prototype first.
